glib/src/subclass/types.rs

```diff
@@ -128,6 +107,103 @@ pub unsafe trait IsImplementable<T: ObjectSubclass>: StaticType {
+        vec![(
+            A::static_type().to_glib(),
+            gobject_ffi::GInterfaceInfo {
+                interface_init: Some(A::interface_init),
+                interface_finalize: None,
+                interface_data: ptr::null_mut(),
+            },
+        )]
+    }
+}
+
+// Generates all the InterfaceList impls for interface_lists of arbitrary sizes based on a list of type
+// parameters like A B C. It would generate the impl then for (A, B) and (A, B, C).
+macro_rules! interface_list_trait(
+    ($name1:ident, $name2: ident, $($name:ident),*) => (
+        interface_list_trait!(__impl $name1, $name2; $($name),*);
+    );
+    (__impl $($name:ident),+; $name1:ident, $($name2:ident),*) => (
+        interface_list_trait_impl!($($name),+);
+        interface_list_trait!(__impl $($name),+ , $name1; $($name2),*);
+    );
+    (__impl $($name:ident),+; $name1:ident) => (
+        interface_list_trait_impl!($($name),+);
+        interface_list_trait_impl!($($name),+, $name1);
+    );
+);
+
+// Generates the impl block for InterfaceList on interface_lists or arbitrary sizes based on its
+// arguments. Takes a list of type parameters as parameters, e.g. A B C
+// and then implements the trait on (A, B, C).
+macro_rules! interface_list_trait_impl(
+    ($($name:ident),+) => (
+        impl<T: ObjectSubclass, $($name: IsImplementable<T>),+> InterfaceList<T> for ( $($name),+ ) {
+            fn iface_infos() -> Vec<(ffi::GType, gobject_ffi::GInterfaceInfo)> {
+                let mut types = Vec::new();
+                interface_list_trait_inner!(types, $($name)+)
+            }
+        }
+    );
+);
+
+// Generates the inner part of the InterfaceList::types() implementation, which will
+// basically look as follows:
+//
+// let mut types = Vec::new();
+//
+// types.push((A::static_type().to_glib(), ...));
+// types.push((B::static_type().to_glib(), ...));
+// [...]
+// types.push((Z::static_type().to_glib(), ...));
+//
+// types
+macro_rules! interface_list_trait_inner(
+    ($types:ident, $head:ident $($id:ident)+) => ({
+        $types.push(
+            (
+                $head::static_type().to_glib(),
+                gobject_ffi::GInterfaceInfo {
+                    interface_init: Some($head::interface_init),
+                    interface_finalize: None,
+                    interface_data: ptr::null_mut(),
+                },
+            )
+        );
+        interface_list_trait_inner!($types, $($id)+)
+    });
+    ($types:ident, $head:ident) => ({
+        $types.push(
+            (
+                $head::static_type().to_glib(),
+                gobject_ffi::GInterfaceInfo {
+                    interface_init: Some($head::interface_init),
+                    interface_finalize: None,
+                    interface_data: ptr::null_mut(),
+                },
+            )
+        );
+        $types
+    });
+);
+
+interface_list_trait!(A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S);
```

Comment:
I don't understand at all what this is about... What's the point of having this implemented randomly?

Reply:
What do you mean with randomly? This is implementing the `InterfaceList` trait on tuples of size 2 to 19

Reply:
Exactly that. Do we really need to go up to 19 for example? Isn't it possible to make implementations on a trait that the tuples could implement?

Reply:
Well we could go up to any other number you would prefer. Implementing it not separately for each tuple size would require https://github.com/rust-lang/rfcs/issues/376 and until then a macro seemed nicer than duplicating this code a dozen times.

Reply:
Oh damn, I was sure it was merged since the time... Then there is no issue.